src/mutable_list.jl

```diff
@@ -0,0 +1,213 @@
+mutable struct ListNode{T}
+    data::T
+    prev::ListNode{T}
+    next::ListNode{T}
+    function ListNode{T}() where T
+        node = new{T}()
+        node.prev = node
+        node.next = node
+        return node
+    end
+    function ListNode{T}(data) where T
+        node = new{T}(data)
+        return node
+    end
+end
+
+mutable struct MutableLinkedList{T}
+    len::Int
+    front::ListNode{T}
+    back::ListNode{T}
+    function MutableLinkedList{T}() where T
+        l = new{T}()
+        l.len = 0
+        l.front = ListNode{T}()
+        l.back = ListNode{T}()
+        l.front.next = l.back
+        l.back.prev = l.front
+        return l
+    end
+end
+
+MutableLinkedList() = MutableLinkedList{Any}()
+
+function MutableLinkedList{T}(elts...) where T
+    l = MutableLinkedList{T}()
+    for elt in elts
+        push!(l, elt)
+    end
+    return l
+end
+
+iterate(l::MutableLinkedList) = l.len == 0 ? nothing : (l.front.next.data, l.front.next.next)
+iterate(l::MutableLinkedList, n::ListNode) = n.next == n ? nothing : (n.data, n.next)
+
+isempty(l::MutableLinkedList) = l.len == 0
+length(l::MutableLinkedList) = l.len
+collect(l::MutableLinkedList{T}) where T = T[x for x in l]
+eltype(l::MutableLinkedList{T}) where T = T
+first(l::MutableLinkedList) = l.front.next.data
+last(l::MutableLinkedList) = l.back.prev.data
+
+==(l1::MutableLinkedList{T}, l2::MutableLinkedList{S}) where {T,S} = false
+
+function ==(l1::MutableLinkedList{T}, l2::MutableLinkedList{T}) where T
+    length(l1) == length(l2) || return false
+    for (i, j) in zip(l1, l2)
+        i == j || return false
+    end
+    return true
+end
+
+function map(f::Function, l::MutableLinkedList{T}) where T
+    l2 = MutableLinkedList{T}()
```

Comment:
This will break if the return type of `f(h)` is not `T`.

You probably want something like (untested):
```julia
    S = promote_type(Base.returntypes(f, (T,))...)
    l2 = MutableLinkedList{S}()
```

Alternatively (or in addition), you could allow the user to specify the desired eltype.

Reply:
I couldn't find anything about a `returntypes` function, but what about something like this?
 ```julia
    S = typeof(f(first(l)))
    l2 = MutableLinkedList{S <: T ? T : S}()
```
It doesn't seem like the most elegant solution, but I'm not sure what a better approach would be (unless it makes more sense to just have the user specify the type).

Reply:
Actually, it was supposed to be `return_types` (with an underscore).  But reading a little further, it's use is supposed to be more targeted.

See https://discourse.julialang.org/t/obtaining-a-functions-output-type/11313/5 for a full discussion on how `map` is recommended to be implemented.

It's a little involved, though I don't think it's too complicated.  If you decide not to implement the full version here, I think what you specified above is fine (it's similar to the first step in the post)--just leave a breadcrumb to the full implementation.

Reply:
So, I think this is basically the approach that is suggested in that discussion:
```julia
function map(f::Base.Callable, l::MutableLinkedList{T}) where T
    if isempty(l) && f isa Function
        S = Core.Compiler.return_type(f, (T,))
        return MutableLinkedList{S}()
    elseif isempty(l) && f isa Type
        return MutableLinkedList{f}()
    else
        S = typeof(f(first(l)))
        l2 = MutableLinkedList{S}()
        for h in l
            el = f(h)
            if el isa S
                push!(l2, el)
            else
                R = typejoin(S, typeof(el))
                l2 = MutableLinkedList{R}(collect(l2)...)
                push!(l2, el)
            end
        end
        return l2
    end
end
```
It seems to work as expected for most inputs, but for something like this:
```julia
function f(x)
    if x % 2 == 0
        return convert(Int8, x)
    else
        return convert(Float16, x)
    end
end
```
`Core.Compiler.return_type`  returns the `Union` of the types rather than the closest common ancestor (as `typejoin` does). For example:
```julia
julia> map(f, MutableLinkedList{Float64}())
MutableLinkedList{Union{Float16, Int8}}()

julia> map(f, MutableLinkedList{Float64}(1:10...))
MutableLinkedList{Real}(1.0, 2, 3.0, 4, 5.0, 6, 7.0, 8, 9.0, 10)
```
Is that something worth worrying about or not really?